syntax = "proto3";
package measurement;

import "google/protobuf/empty.proto";
import "room.proto";

option go_package = ".;pb"; 

// 測定値の種類
enum MeasurementType {
    TEMPERATURE       = 0;  // 温度
    HUMIDITY          = 1;  // 湿度
    AIR_PRESSURE      = 2;  // 気圧
    CO2_CONCENTRATION = 3;  // CO2濃度
}

// 測定値メッセージの定義
message Measurement {
    room.RoomNumbers room_number = 1;  // 部屋番号
    MeasurementType type = 2;  // 測定値の種類
    float value = 3;  // 実際の測定値
    int64 datetime = 4;  // 測定した時間(unix_time)
}

message MeasurementRequest {
    MeasurementType type = 1;
}

message MeasurementResponse {
    Measurement measurement = 1;
    string message = 2;
    bool error = 3;
}

message InsertResponse {
    reserved 1;
    string message = 2;
    bool error = 3;
}

message AvailableTypes {
    repeated MeasurementType available_types = 1;
}

service MeasurementRecorder {
    // Administrator -> Servant
    // ある特定の種類のセンサーの値を返す
    rpc Get (MeasurementRequest) returns (MeasurementResponse) {}
    // 利用可能なセンサーの種類を返す
    rpc GetAvailableTypes (google.protobuf.Empty) returns (AvailableTypes) {}
}

service MeasurementStore{
    // Administrator <- Servant
    // ある特定の種類のセンサーの値をDBに登録する
    rpc InsertData (Measurement) returns (InsertResponse) {}
}